README
tumble: build a PDF file from image files
Copyright 2003-2017 Eric Smith <spacewar@gmail.com>

Tumble is a utility to construct PDF files from one or more image
files.  Supported input image file formats are JPEG, and black and
white TIFF (single- or multi-page).  Black and white images will be
encoded in the PDF output using lossless Group 4 fax compression
(ITU-T recommendation T.6).  This provides a very good compression
ratio for text and line art.  JPEG images will be preserved with the
original coding.

The current version of Tumble will only work on little-endian systems,
such as x86, VAX, and Alpha.  The byte order dependencies will be fixed
in a later release.

The input and output files can be specified on the command line.
Alternatively, a control file, typically with a ".tum" suffix, may be
used which allows for more control over the files and options.

The general command line syntax is:

    tumble [options] <input>... -o <output.pdf>

The options in this mode are:

    -v        verbose
    -b <fmt>  create bookmarks

If the "-b" option is given, bookmarks will be created using the
format string, which may contain arbitrary text and/or the following
format specifiers:

    %F  file name, sans suffix  e.g., "foo.tif" will just appear as "foo"
    %p  page number of input file, useful for multipage TIFF files

There is currently no documentation for the control file syntax, as it
is still being refined, and many of the options planned for use in
control files are not yet fully implemented.  Features that will be
available in control files include:

    image resolution specification - useful for input files with
        unspecified resolution, or to override
    image rotation, in units of 90 degrees
    image cropping
    grouping to allow different specifications for individual pages
        or groups of consecutive pages (e.g., chapters)
    ability to control operations independently on even and odd pages
        (e.g., different rotation or cropping)
    more flexible bookmark control
    page labels, so that the PDF file can present the actual document
        page numbers such as xvii or 3-27


This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License version 2 as
published by the Free Software Foundation.  Note that permission is
not granted to redistribute this program under the terms of any other
version of the General Public License.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111 USA


CHANGELOG.md
# Change Log

## [0.36] - 2017-12-27
### Minor updates, no significant code changes.
- Minor changes to avoid a few compile warnings.
- Updates to email address, URL, copyright notices.

## [0.35] - 2016-06-23
### Added CHANGELOG.md	
	
## [0.34] - 2016-06-23
### Many fixes and enhancements from Palomar Corporation
- Modify "bitblt.h" to undefine BITS_PER_WORD, which is defined in the
  netpbm header "pm_config.h", so that the tumble use takes precedence.
- Fix a bug in "bitblt_write_g4" (bitblt_g4.c).  Tumble would crash in
  "g4_get_pixel" with bad pointer access when trying to encode a TIFF image
  that was 14400 x 6100 (24" x 11").  "g4_get_pixel" was called with index
  14400, which accesses the first byte beyond the allocated array.
- "bitblt_write_g4" is called to encode and write a TIFF bitmap to the PDF
  file.  It allocates and zeros a temporary buffer to hold a scan line and
  then calls "g4_encode_row" for each line.  "g4_find_pixel" is called to
  find the first 1 bit in the temporary buffer.  Because the buffer has been
  zeroed, "g4_find_pixel" returned the image width (14400) when the
  specified pixel could not be found in scan line.  A subsequent call to
  "g4_get_pixel" specified this pixel number (14400), which accessed memory
  beyond the allocation.

  The crash wasn't repeatable.  To occur, the image width had to be a
  multiple of 32 (else the last pixel + 1 would still be within the final
  word), and the allocation had to end exactly on a page boundary (so the
  first byte beyond the allocation was read-protected).
	
  Fix is to pad the temporary buffer with an extra word if the width is an
  exact multiple of 32.  Note that access beyond the source bitmap is
  prohibited by the "if (a0 >= width) break;" test prior to the
  "g4_get_pixel" call.

- Modify the makefile to add PNG and blank-page support files.

- Modify "parser.y" to enable push and pop of input contexts.  This allows
  "{" and "}" to create a local context for modifiers, such as TRANSPARENT.
  Input modifiers specified outside of a local context become global for
  subsequent input clauses.
- Modify "pdf.c" to add creation and modification dates to the PDF
  properties.  The "CreationDate" and "ModDate" values are set to the time
  that the PDF was created.
- Modify "pdf.h" to change the page size limit from 45 inches (Reader 3.x)
  to 200 inches (Reader 4.x and later), to add structures and parameters to
  support the new features added to the parser, and to add the PDF
  "producer" definition.
- Modify "pdf_g4.c" to add support for color-mapping bilevel images.
  Mapping allows a black and white image to be drawn as any two specified
  colors.  Useful for pages that are black-on-color or white-on-color. The
  color mapping syntax is:

       COLORMAP (<red> <green> <blue>) , (<red> <green> <blue>) ;

  ...where the first triplet specifies the color to be displayed for black
  and the second triplet specifies the color to be displayed for white.

  Note that a two-color PNG is about twice the size of a Group 4 TIFF.

  Implementation note: the colormap is added to the PDF as a string resource
  and an indirect reference is made from the XObject.  Tumble should keep
  each colormap only once in the file by keeping a global list of the
  string resources created.  It doesn't.  However, if the colormap used on
  an image is identical to the one used on the previous image, it will reuse
  the string resource.  This covers the case of a section of colored pages.
  However, a discontiguous run of colored pages, e.g., the front and back
  covers, will store the colormap twice.

- Modify "pdf_g4.c", "pdf_jpeg.c", "tumble.c", "tumble_jpeg.c", and
  "tumble_pbm.c" to support color key masking (using the PDF Mask operator).
  The syntax is:

       TRANSPARENT ( <gray_range> ) ;

  or:

       TRANSPARENT ( <red_range> <green_range> <blue_range> ) ;

  ...as a modifier to an input statement.  <range> may be either a single
  integer or a low..high range.

  The range values must be compatible with the color space of the image.
  For full-color images, the ranges would be 0..255.  For grayscale images,
  the range would be 0..255.  For paletted images, the range  is expressed
  in palette entry numbers, which is much less likely to be useful than a
  single palette value.

- Modify "pdf_g4.c", "tumble_input.h", and "tumble_tiff.c" to add handling
  of TIFF images with min-is-black photometric interpretation.

- Modify "pdf_g4.c," "pdf_jpeg.c," and "pdf_png.c" to fix bugs in the
  callbacks within these files.  "pdf_write_stream" was called to write "Do"
  commands with the XObject names without escaping restricted characters in
  the name, which had been escaped when the corresponding directory was
  written.  This led to "Undefined XObject resource" errors when more than
  26 images were overlaid on a page ("pdf_new_XObject" generates names of
  the form "ImA", "ImB", etc., and the 27th name was "Im[", which must be
  escaped, as "[" is a restricted character; the dictionary name was being
  escaped to "Im#5b", but the "Do" command name was not).

- Modify "pdf_page_label.c" to fix a bug wherein a page label specifying
  prefix without a style (e.g., LABEL <prefix>) produces bad PDF (no labels
  are displayed).  Should output "/P <prefix>" but instead was outputting
  "/S /P <prefix>".

- Add new files "pdf_png.c" and "tumble_png.c" and modify "tumble_input.h"
  to support PNG images (code from Daniel Glöckner via the tumble mailing
  list).

- Change PDF string handling in "pdf_prim.c" and "pdf_prim.h" to allow
  embedded nulls by storing strings as character arrays plus length words.

- Modify "parser.y", "scanner.l", "semantics.c", and "semantics.h" to
  support blank pages, overlay images, color mapping, color-key masking, and
  push/pop of input contexts.

- Modify "tumble.c", "tumble_jpeg.c", and "tumble_pbm.c" to support overlay

pdf.c
/*
 * tumble: build a PDF file from image files
 *
 * PDF routines
 * Copyright 2001, 2002, 2003, 2017 Eric Smith <spacewar@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.  Note that permission is
 * not granted to redistribute this program under the terms of any
 * other version of the General Public License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111 USA
 *
 *  2005-05-03 [JDB] Add CreationDate and ModDate PDF headers.
 *  2014-02-18 [JDB] Use PDF_PRODUCER definition.
 */


#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>


#include "bitblt.h"
#include "pdf.h"
#include "pdf_util.h"
#include "pdf_prim.h"
#include "pdf_private.h"
#include "pdf_name_tree.h"


static void pdf_set_info (pdf_file_handle pdf_file, char *key, char *val)
{
  if (! pdf_file->info)
    pdf_file->info = pdf_new_ind_ref (pdf_file, pdf_new_obj (PT_DICTIONARY));

  pdf_set_dict_entry (pdf_file->info, key, pdf_new_string (val));
}


void pdf_init (void)
{
}


struct pdf_pages *pdf_new_pages (pdf_file_handle pdf_file)
{
  struct pdf_pages *pages = pdf_calloc (1, sizeof (struct pdf_pages));

  pages->kids = pdf_new_obj (PT_ARRAY);
  /* The PDF 1.0 spec doesn't say that kids can't be an indirect object,
     but Acrobat 4.0 fails to optimize files if it is. */

  pages->count = pdf_new_integer (0);
  pages->pages_dict = pdf_new_ind_ref (pdf_file, pdf_new_obj (PT_DICTIONARY));
  pdf_set_dict_entry (pages->pages_dict, "Type", pdf_new_name ("Pages"));
  pdf_set_dict_entry (pages->pages_dict, "Kids", pages->kids);
  pdf_set_dict_entry (pages->pages_dict, "Count", pages->count);
  return (pages);
}


pdf_file_handle pdf_create (char *filename)
{
  pdf_file_handle pdf_file;
  time_t current_time, adjusted_time;
  struct tm *time_and_date;
  int gmt_diff;
  char tm_string[18], gmt_string[24];
  const char tm_format[] = "D:%Y%m%d%H%M%S";

  pdf_file = pdf_calloc (1, sizeof (struct pdf_file));

  pdf_file->f = fopen (filename, "wb");
  if (! pdf_file->f)
    {
      pdf_fatal ("error opening output file\n");
    }

  pdf_file->root = pdf_new_pages (pdf_file);

  pdf_file->catalog = pdf_new_ind_ref (pdf_file, pdf_new_obj (PT_DICTIONARY));
  pdf_set_dict_entry (pdf_file->catalog, "Type", pdf_new_name ("Catalog"));
  pdf_set_dict_entry (pdf_file->catalog, "Pages", pdf_file->root->pages_dict);
  /* Outlines dictionary will be created later if needed */
  pdf_set_dict_entry (pdf_file->catalog, "PageLayout", pdf_new_name ("SinglePage"));

  pdf_file->info    = pdf_new_ind_ref (pdf_file, pdf_new_obj (PT_DICTIONARY));
  pdf_set_info (pdf_file, "Producer", PDF_PRODUCER);

  /* Generate CreationDate and ModDate */

  current_time = time (NULL);
  time_and_date = gmtime (&current_time);
  adjusted_time = mktime (time_and_date);
  gmt_diff = (int) difftime (current_time, adjusted_time);

  time_and_date = localtime (&current_time);

  if (strftime (tm_string, sizeof (tm_string), tm_format, time_and_date))
    {
      sprintf (gmt_string, "%s%+03d'%02d'", tm_string, gmt_diff / 3600, gmt_diff % 60);
      pdf_set_info (pdf_file, "CreationDate", gmt_string);
      pdf_set_info (pdf_file, "ModDate", gmt_string);
    }

  pdf_file->trailer_dict = pdf_new_obj (PT_DICTIONARY);
  /* Size key will be added later */
  pdf_set_dict_entry (pdf_file->trailer_dict, "Root", pdf_file->catalog);
  pdf_set_dict_entry (pdf_file->trailer_dict, "Info", pdf_file->info);

  /* write file header */
  fprintf (pdf_file->f, "%%PDF-1.3\r\n");

  /* write comment containing 8-bit chars as a hint that the file is binary */
  /* PDF 1.4 spec, section 3.4.1 */
  fprintf (pdf_file->f, "%%\342\343\317\323\r\n");

  return (pdf_file);
}


void pdf_close (pdf_file_handle pdf_file, int page_mode)
{
  char *page_mode_string;

  page_mode_string = "UseNone";

  switch (page_mode)
    {
    case PDF_PAGE_MODE_USE_NONE:
      break;
    case PDF_PAGE_MODE_USE_OUTLINES:
      if (pdf_file->outline_root)
	page_mode_string = "UseOutlines";
      break;
    case PDF_PAGE_MODE_USE_THUMBS:
      page_mode_string = "UseThumbs";
      break;
    default:
      pdf_fatal ("invalid page mode\n");
    }

  pdf_set_dict_entry (pdf_file->catalog,
		      "PageMode",
		      pdf_new_name (page_mode_string));

  /* finalize trees, object numbers aren't allocated until this step */
  pdf_finalize_name_trees (pdf_file);

  /* add the page label number tree, if it exists, to the catalog */
  if (pdf_file->page_label_tree)
    pdf_set_dict_entry (pdf_file->catalog,
			"PageLabels",
			pdf_file->page_label_tree->root->dict);

  /* write body */
  pdf_write_all_ind_obj (pdf_file);

  /* write cross reference table and get maximum object number */
  pdf_set_dict_entry (pdf_file->trailer_dict, "Size", pdf_new_integer (pdf_write_xref (pdf_file)));

  /* write trailer */
  fprintf (pdf_file->f, "trailer\r\n");
  pdf_write_obj (pdf_file, pdf_file->trailer_dict);
  fprintf (pdf_file->f, "startxref\r\n");
  fprintf (pdf_file->f, "%ld\r\n", pdf_file->xref_offset);
  fprintf (pdf_file->f, "%%%%EOF\r\n");

  fclose (pdf_file->f);
  /* should free stuff here */
}


void pdf_set_author   (pdf_file_handle pdf_file, char *author)
{
  pdf_set_info (pdf_file, "Author", author);
}

void pdf_set_creator  (pdf_file_handle pdf_file, char *creator)
{
  pdf_set_info (pdf_file, "Creator", creator);
}

void pdf_set_producer  (pdf_file_handle pdf_file, char *producer)
{
  pdf_set_info (pdf_file, "Producer", producer);
}

void pdf_set_title    (pdf_file_handle pdf_file, char *title)
{
  pdf_set_info (pdf_file, "Title", title);
}

void pdf_set_subject  (pdf_file_handle pdf_file, char *subject)
{
  pdf_set_info (pdf_file, "Subject", subject);
}

void pdf_set_keywords (pdf_file_handle pdf_file, char *keywords)
{
  pdf_set_info (pdf_file, "Keywords", keywords);
}


pdf_page_handle pdf_new_page (pdf_file_handle pdf_file,
			      double width,
			      double height)
{
  pdf_page_handle page = pdf_calloc (1, sizeof (struct pdf_page));

  page->pdf_file = pdf_file;

  page->media_box = pdf_new_obj (PT_ARRAY);
  pdf_add_array_elem (page->media_box, pdf_new_real (0));
  pdf_add_array_elem (page->media_box, pdf_new_real (0));
  pdf_add_array_elem (page->media_box, pdf_new_real (width));
  pdf_add_array_elem (page->media_box, pdf_new_real (height));

  page->procset = pdf_new_obj (PT_ARRAY);
  pdf_add_array_elem_unique (page->procset, pdf_new_name ("PDF"));

  page->resources = pdf_new_obj (PT_DICTIONARY);
  pdf_set_dict_entry (page->resources, "ProcSet", page->procset);

  page->page_dict = pdf_new_ind_ref (pdf_file, pdf_new_obj (PT_DICTIONARY));
  pdf_set_dict_entry (page->page_dict, "Type", pdf_new_name ("Page"));
  pdf_set_dict_entry (page->page_dict, "MediaBox", page->media_box);
  pdf_set_dict_entry (page->page_dict, "Resources", page->resources);

  if (pdf_get_integer (pdf_file->root->count) == 0)
    {
      struct pdf_obj *dest_array = pdf_new_obj (PT_ARRAY);
      pdf_add_array_elem (dest_array, page->page_dict);
      pdf_add_array_elem (dest_array, pdf_new_name ("Fit"));
      pdf_set_dict_entry (pdf_file->catalog, "OpenAction", dest_array);
    }

  /* $$$ currently only support a single-level pages tree */
  pdf_set_dict_entry (page->page_dict, "Parent", pdf_file->root->pages_dict);
  pdf_add_array_elem (pdf_file->root->kids, page->page_dict);
  pdf_set_integer (pdf_file->root->count,
		   pdf_get_integer (pdf_file->root->count) + 1);

  page->last_XObject_name = '@';  /* first name will be "ImA" */

  return (page);
}

void pdf_close_page (pdf_page_handle pdf_page)
{
}


void pdf_set_page_number (pdf_page_handle pdf_page, char *page_number)
{
}

